### PR TITLE
[JENKINS-20046] - Do not query queue dispatchers from UI

### PR DESCRIPTION
**Pull request by Jimilian**
During work on https://github.com/jenkinsci/gerrit-trigger-plugin/pull/325 I was surprised that Jenkins calls `canRun` for `QueueTaskDispatcher.all()` during `getWhy` every time User opens the UI... So, I decided to fix this - maybe it will help a little bit for tickets I mentioned below:

See [JENKINS-27650](https://issues.jenkins-ci.org/browse/JENKINS-27650) and [JENKINS-20046](https://issues.jenkins-ci.org/browse/JENKINS-20046).

### Proposed changelog entries

* Do not query queue dispatchers from UI

### Submitter checklist

- [-] JIRA issue is well described
- [+] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [+] Appropriate autotests or explanation to why this change has no tests

I did not write any new tests because this PR should not change any behaviour in the system.

### CODE REVIEW

File: core/src/main/java/hudson/model/Queue.java [hunk 1]
```diff
@@ -1175,17 +1176,26 @@ public boolean contains(Task t) {
...
-        if (i.task.isBuildBlocked() || !canRun(i.task.getResourceList()))
-            return true;
+    @Nullable
+    private CauseOfBlockage getCauseOfBlockage(Item i) {
+        if (i.task.isBuildBlocked())
+            return i.task.getCauseOfBlockage();
+
+        if (!canRun(i.task.getResourceList())) {
+            // TODO remove
+            return CauseOfBlockage.createNeedsMoreExecutor(Messages._Queue_WaitingForNextAvailableExecutor());
```
Comment (Jimilian) [inline core/src/main/java/hudson/model/Queue.java]: I did not find what to put here, any suggestions?

File: core/src/main/java/hudson/model/Queue.java [hunk 2]
```diff
@@ -1175,17 +1176,26 @@ public boolean contains(Task t) {
     /**
      * Checks if the given item should be prevented from entering into the {@link #buildables} state
      * and instead stay in the {@link #blockedProjects} state.
+     *
+     * @return the reason of blockage if it exists null otherwise.
      */
-    private boolean isBuildBlocked(Item i) {
-        if (i.task.isBuildBlocked() || !canRun(i.task.getResourceList()))
-            return true;
+    @Nullable
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: `@CheckForNull`. Nullable does not strictly require the null check

File: core/src/main/java/hudson/model/Queue.java [hunk 3]
```diff
@@ -2443,12 +2458,20 @@ protected NotWaitingItem(NotWaitingItem ni) {
      * {@link Item} in the {@link Queue#blockedProjects} stage.
      */
     public final class BlockedItem extends NotWaitingItem {
-        public BlockedItem(WaitingItem wi) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: It was a public class. The old constructor has to be retained. `causeOfBlockage` is nullable, hence it should not be a problem

File: core/src/main/java/hudson/model/Queue.java [hunk 4]
```diff
@@ -2459,10 +2482,8 @@ public CauseOfBlockage getCauseOfBlockage() {
                 return CauseOfBlockage.fromMessage(Messages._Queue_BlockedBy(r.getDisplayName()));
             }
 
-            for (QueueTaskDispatcher d : QueueTaskDispatcher.all()) {
-                CauseOfBlockage cause = d.canRun(this);
-                if (cause != null)
-                    return cause;
+            if (causeOfBlockage != null) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: Fine for me. Probably it could be checked even before calling `getBlockingActivity()` to speedup it even more. Though it needs more feedback from @jenkinsci/code-reviewers

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
I like the improvement, it would be useful and compliant with the Queue snapshot idea.

The compatibility issue needs to be fixed though

File: core/src/main/java/hudson/model/Queue.java (hunk 3 above)
Comment (Jimilian) [reply core/src/main/java/hudson/model/Queue.java]: @oleg-nenashev Can I mark old constructors as `@Deprecated` in this case?

Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/Queue.java]: works for me

File: core/src/main/java/hudson/model/Queue.java (hunk 4 above)
Comment (Jimilian) [reply core/src/main/java/hudson/model/Queue.java]: Can I move it to getCauseOfBlockage as well? In this case the load from it would be consistent (at least :) ). But I'm not sure about it... From my personal perspective it would be better just remove it :))

**Comment by Jimilian** [comment]
@jglick, @stephenc, @oleg-nenashev, please, review.

**Comment by oleg-nenashev** [comment]
> Can I move it to getCauseOfBlockage as well? In this case the load from it would be consistent (at least :) ). But I'm not sure about it... From my personal perspective it would be better just to remove it from UI :))

I'd guess it's a good subject for the follow-up discussion

File: core/src/main/java/hudson/model/Queue.java [hunk 5]
```diff
@@ -1487,6 +1519,8 @@ public void maintain() {
                             // determine if they are blocked by the lucky winner
                             updateSnapshot();
                         }
+                    } else {
+                        p.setCauseOfBlockage(causeOfBlockage);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: if ``allowNewBuildableTask` returns false, a non-null value may be overridden by a null one here. Is it "as designed"?

Comment (Jimilian) [reply core/src/main/java/hudson/model/Queue.java]: Good catch. Moved `allowNewBuildableTask` to  `getCauseOfBlockage` as well.

**Comment by Jimilian** [comment]
@daniel-beck, can you estimate when will you have time for review, please? I'm just curious is there any change to get it merge any time soon or not.

**Comment by daniel-beck** [comment]
@Jimilian Unsure when I'll have the time. Please note that I use self-requested reviews for things I'd just like to look at, it's different (to me) from assignment. Nobody should assume I'm blocking this in any way.

**Review by oleg-nenashev — APPROVED** [review]
Looks good to me

File: core/src/main/java/hudson/model/Queue.java [hunk 6]
```diff
@@ -1175,28 +1175,63 @@ public boolean contains(Task t) {
...
      * Checks if the given item should be prevented from entering into the {@link #buildables} state
      * and instead stay in the {@link #blockedProjects} state.
+     *
+     * @return the reason of blockage if it exists null otherwise.
      */
-    private boolean isBuildBlocked(Item i) {
-        if (i.task.isBuildBlocked() || !canRun(i.task.getResourceList()))
-            return true;
+    @CheckForNull
+    private CauseOfBlockage getCauseOfBlockageForItem(Item i) {
```
Comment (KostyaSha) [inline core/src/main/java/hudson/model/Queue.java]: sounds overused for non static method, why not just?
`getCauseOfBlockage(Item i)`
`getCauseOfBlockage(Task i)`

Comment (Jimilian) [reply core/src/main/java/hudson/model/Queue.java]: This file contains already several `getCauseOfBlockage` methods, so, this name is just tradeoff to reduce cognitive complexity of this file :)

Comment (KostyaSha) [reply core/src/main/java/hudson/model/Queue.java]: not a problem while they have argument that are not inherited by each other =)

**Comment by oleg-nenashev** [comment]
@KostyaSha are you fine with that?

**Comment by KostyaSha** [comment]
Well, i'm fine with anything, just thing that specifying argument type in method name makes no sense when it possible not to have it.

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers I am going to merge it today if there is no negative feedback

File: core/src/main/java/hudson/model/Queue.java [hunk 7]
```diff
@@ -2443,29 +2485,45 @@ protected NotWaitingItem(NotWaitingItem ni) {
...
 
+        /**
+         * @deprecated Use {@link #BlockedItem(NotWaitingItem, CauseOfBlockage)} instead
+         */
+        @Deprecated
         public BlockedItem(NotWaitingItem ni) {
+            this(ni, null);
+        }
+
+        public BlockedItem(WaitingItem wi, CauseOfBlockage causeOfBlockage) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: Why would this be `public`? It is only called from within the same source file AFAIK.

File: core/src/main/java/hudson/model/Queue.java [hunk 8]
```diff
@@ -2443,29 +2485,45 @@ protected NotWaitingItem(NotWaitingItem ni) {
...
         }
 
-        public CauseOfBlockage getCauseOfBlockage() {
-            ResourceActivity r = getBlockingActivity(task);
-            if (r != null) {
-                if (r == task) // blocked by itself, meaning another build is in progress
-                    return CauseOfBlockage.fromMessage(Messages._Queue_InProgress());
-                return CauseOfBlockage.fromMessage(Messages._Queue_BlockedBy(r.getDisplayName()));
-            }
+        public void setCauseOfBlockage(CauseOfBlockage causeOfBlockage) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: Ditto. Why would this need to be a `public` API? What plugin would be calling it and why?

**Review by jglick — COMMENTED** [review]
For the future, rather than merely mentioning someone you wish to have a review from, actually use the GitHub gesture to request them as a reviewer.

**Comment by jglick** [comment]
> this PR should not change any behaviour in the system

That does not seem to be true. IIUC the expectation is that after certain pages are loaded, or APIs are called, or whatever, previously the `canRun` method would be called anew for a test extension; and now it will not. That is testable.

File: core/src/main/java/hudson/model/Queue.java [hunk 9]
```diff
@@ -1802,6 +1841,9 @@ public Api getApi() {
          */
         @CheckForNull
         default CauseOfBlockage getCauseOfBlockage() {
+            if (isBuildBlocked()) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: [JENKINS-47517](https://issues.jenkins-ci.org/browse/JENKINS-47517)

File: core/src/main/java/hudson/model/Queue.java [hunk 10]
```diff
@@ -1191,6 +1191,20 @@ private CauseOfBlockage getCauseOfBlockage(Item i) {
                 return causeOfBlockage;
         }
 
+        if(!(i instanceof BuildableItem)) {
+            // Make sure we don't queue two tasks of the same project to be built
+            // unless that project allows concurrent builds. Once item is buildable it's ok.
+            //
+            // This check should never pass. And must be remove once we can completely rely on `getCauseOfBlockage`.
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java:1198]: Unclear whether we can now “completely rely on `getCauseOfBlockage`”. But this code just seems to be a reworking of some much older logic which I find hard to follow, from 84bca061dc4f330d17999bb15ee4b40cac227fd9 & 3216dface66707a1b069e51912513ef02beac79c. Perhaps this has been obsolete since bc66388f8550e748f898164711e352e6eb5da996? System dates to c04e63d9f0c0b1a3b0683e64838c685cdddcd265.
